File: server/src/main/java/com/yoedu/job_board_platform/models/Job.java
package com.yoedu.job_board_platform.models;

import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "jobs")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EntityListeners(AuditingEntityListener.class)
/**
 * Đại diện cho một tin tuyển dụng trên nền tảng.
 * Chứa thông tin chi tiết về vị trí, mức lương, địa điểm, loại hình làm việc,
 * cấp độ kinh nghiệm và trạng thái của bài đăng.
 */
public class Job {
    @Id
    @Column(columnDefinition = "uuid")
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "company_id", nullable = false)
    private Company company;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "category_id", nullable = false)
    private JobCategory category;

    @Column(nullable = false, length = 255)
    private String title;

    @Column(nullable = false, unique = true)
    private String slug;

    @Column(columnDefinition = "text", nullable = false)
    private String description;

    @Column(columnDefinition = "text")
    private String requirements;

    @Column(columnDefinition = "text")
    private String benefits;

    @Column(name = "number_of_openings")
    @Builder.Default
    private Integer numberOfOpenings = 1;

    @Column(name = "salary_min", precision = 12, scale = 2)
    private BigDecimal salaryMin;

    @Column(name = "salary_max", precision = 12, scale = 2)
    private BigDecimal salaryMax;

    @Column(length = 10)
    @Builder.Default
    private String currency = "VND";

    private String location;

    @Enumerated(EnumType.STRING)
    @Column(name = "location_types", nullable = false)
    private LocationTypes locationTypes;

    @Enumerated(EnumType.STRING)
    @Column(name = "employment_type", nullable = false)
    private EmploymentType employmentType;

    @Enumerated(EnumType.STRING)
    @Column(name = "experience_level", nullable = false)
    private ExperienceLevel experienceLevel;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    @Builder.Default
    private JobStatus status = JobStatus.DRAFT;

    @Column(name = "posted_date")
    private OffsetDateTime postedDate;

    @Column(name = "expiration_date")
    private OffsetDateTime expirationDate;

    @Column(name = "rejection_reason", columnDefinition = "text")
    private String rejectionReason;

    @Column(name = "created_at", nullable = false)
    @CreatedDate
    private OffsetDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    @LastModifiedDate
    private OffsetDateTime updatedAt;

    public boolean hasCategoriesChanged(Integer newCategoryId) {
        return newCategoryId != null && !newCategoryId.equals(this.getCategory().getId());
    }
}

File: client/src/lib/schemas/job.ts
import { EmploymentTypes, ExperienceLevels, LocationTypes } from "@/types/job";
import { z } from "zod";

export const jobSchema = z
	.object({
		title: z.string().min(1, "Tiêu đề không được để trống").max(255, "Tiêu đề tối đa 255 ký tự"),
		description: z.string().min(1, "Mô tả không được để trống"),
		requirements: z.string().min(1, "Yêu cầu ứng viên không được để trống"),
		benefits: z.string().min(1, "Quyền lợi không được để trống"),
		categoryId: z.number({
			error: "Vui lòng chọn ngành nghề",
		}),
		numberOfOpenings: z.number().int().min(1).default(1).optional(),
		salaryMin: z.number().min(0, "Lương không được âm").optional().nullable(),
		salaryMax: z.number().min(0, "Lương không được âm").optional().nullable(),
		currency: z.string().default("VND"),
		location: z.string().default(""),
		locationTypes: z.enum(LocationTypes, { error: "Vui lòng chọn hình thức làm việc" }),
		employmentType: z.enum(EmploymentTypes, { error: "Vui lòng chọn loại hình" }),
		experienceLevel: z.enum(ExperienceLevels, { error: "Vui lòng chọn cấp bậc kinh nghiệm" }),
		skillIds: z.array(z.number()).optional().default([]),
	})
	.refine(
		(data) => {
			if (data.salaryMin != null && data.salaryMax != null) {
				return data.salaryMax >= data.salaryMin;
			}
			return true;
		},
		{ message: "Lương tối đa phải lớn hơn hoặc bằng lương tối thiểu", path: ["salaryMax"] },
	);

export type JobFormData = z.input<typeof jobSchema>;
